### PR TITLE
Add operation interfaces for upcoming relational abstraction support

### PR DESCRIPTION
**Pull request by jnmt**
This PR introduces operation interfaces to re-use the same logic for the new relational abstraction operation (i.e., `RelationalScan` with arbitrary conditions, and it's introduced by another PR) while keeping backward compatibility.

Also, operations in the relational abstraction don't have to have partition keys and clustering keys. So, I'm adding a more basic operation called `BaseOperation` than the current `Operation` class.

In short, what we would like to achieve here is that both `Scan` and `RelationalScan` implements `ScanInterface` but 
those two have different origins, i.e., `Operation` (with keys) and `BaseOperation` (without keys).

### CODE REVIEW

File: core/src/main/java/com/scalar/db/api/SelectionInterface.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package com.scalar.db.api;
+
+import java.util.List;
+
+/** An interface for selection operations. */
+public interface SelectionInterface extends BaseOperationInterface {
+
+  /**
+   * Returns the table name for this operation
```
Comment (komamitsu) [inline core/src/main/java/com/scalar/db/api/SelectionInterface.java:9]: "column names"?

File: core/src/main/java/com/scalar/db/api/Operation.java [hunk 2]
```diff
@@ -9,72 +9,59 @@
...
   /**
    * Returns the namespace for this operation
    *
    * @return an {@code Optional} with the returned namespace
    */
+  @Override
   @Nonnull
   public Optional<String> forNamespace() {
-    return namespace;
+    return super.forNamespace();
```
Comment (komamitsu) [inline core/src/main/java/com/scalar/db/api/Operation.java:42]: I think you can remove these verbose overridden methods as the superclass's methods are called

File: core/src/main/java/com/scalar/db/storage/jdbc/JdbcDatabase.java [hunk 3]
```diff
@@ -93,7 +93,7 @@ public Optional<Result> get(Get get) throws ExecutionException {
 
   @Override
   public Scanner scan(Scan scan) throws ExecutionException {
-    scan = copyAndSetTargetToIfNot(scan);
+    scan = (Scan) copyAndSetTargetToIfNot(scan);
```
Comment (komamitsu) [inline core/src/main/java/com/scalar/db/storage/jdbc/JdbcDatabase.java:96]: I guess that the caller of `copyAndSetTargetToIfNot(ScanInterface)` expects the same return type as the parameter type (e.g. expects `Scan` when passing `Scan` instance.) If so, how about keeping the existing `Scan copyAndSetTargetToIfNot(Scan)` and adding `RelationalScan copyAndSetTargetToIfNot(RelationalScan)` later so that we can remove this class cast?

Comment (jnmt) [reply core/src/main/java/com/scalar/db/storage/jdbc/JdbcDatabase.java:96]: I'm not sure why I chose the cast rather than adding the relational one, but now I think your suggestion is far better. I will fix it (including other comments). Thank you!

File: core/src/main/java/com/scalar/db/api/SelectionInterface.java (hunk 1 above)
Comment (jnmt) [reply core/src/main/java/com/scalar/db/api/SelectionInterface.java]: Good catch! Thank you! Fixed in d6da904.

File: core/src/main/java/com/scalar/db/storage/jdbc/JdbcDatabase.java (hunk 3 above)
Comment (jnmt) [reply core/src/main/java/com/scalar/db/storage/jdbc/JdbcDatabase.java]: Fixed in d6da904. Thank you!

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java [hunk 4]
```diff
@@ -177,8 +178,8 @@ private TableMetadata getTableMetadata(Scan scan) throws ExecutionException {
     }
   }
 
-  public Optional<List<Key>> get(Scan scan) {
-    if (isWriteSetOverlappedWith(scan)) {
+  public Optional<List<Key>> get(ScanInterface scan) {
+    if (scan instanceof Scan && isWriteSetOverlappedWith((Scan) scan)) {
```
Comment (komamitsu) [inline core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java:182]: This check (or an equivalent) won't be needed for RelationalScan?

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java [hunk 5]
```diff
@@ -318,14 +319,14 @@ void toSerializableWithExtraRead(DistributedStorage storage)
...
-    for (Map.Entry<Scan, List<Key>> entry : scanSet.entrySet()) {
+    for (Map.Entry<ScanInterface, List<Key>> entry : scanSet.entrySet()) {
       tasks.add(
           () -> {
             Map<Key, TransactionResult> currentReadMap = new HashMap<>();
             Set<Key> validatedReadSet = new HashSet<>();
             Scanner scanner = null;
             try {
-              Scan scan = entry.getKey();
+              Scan scan = (Scan) entry.getKey();
```
Comment (komamitsu) [inline core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java:329]: What will this code be like when RelationalScan is introduced?

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java (hunk 4 above)
Comment (jnmt) [reply core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java:182]: Good question. A check for preventing scan-after-write is necessary for RelationalScan too. But it's done at a bit different timing, and the PR will be proposed later.

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java (hunk 5 above)
Comment (jnmt) [reply core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java:329]: It's like the following, and `rebuildScan` will create another Scan operation using a builder pattern.

```java
              ScanInterface scan = entry.getKey();
              if (scan instanceof RelationalScan) {
                scanner =
                    storage.scan(
                        rebuildRelationalScan((RelationalScan) scan, getTableMetadata(scan)));
              } else {
                scanner = storage.scan(rebuildScan((Scan) scan, getTableMetadata(scan)));
              }
```

Comment (komamitsu) [reply core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java:329]: Thanks! It would be great if we have 2 overloaded `rebuildScan(Scan)` and `rebuildScan(RelationalScan)` as we can simply write it like `scanner = storage.scan(rebuildScan(scan, getTableMetadata(scan)))`

Comment (jnmt) [reply core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java:329]: Even if we have 2 overloaded `rebuildScan`, I think we still need a cast and dispatch here since `scan` here is `ScanInterface`. Am I understanding correctly?

**Review by komamitsu — APPROVED** [review]
LGTM! 👍

**Comment by jnmt** [comment]
As a result of a careful discussion with Toshi, we decided to explore another option, extending `Scan` instead of introducing the interface changes. This is mainly because we have several similar interface and abstraction issues (e.g., ScanAll), and they cannot be solved without breaking backward compatibilities if we try to keep smart and elegant interface designs (e.g., not like using XyzInterface). So, anyway, let me make this PR back to the draft (maybe I will create another PR).

@komamitsu Sorry for taking your time for the review.

**Comment by jnmt** [comment]
I will close this draft PR since I submit another PR (#889) with a new solution.
